cranelift/codegen/src/egraph/node.rs

```diff
@@ -0,0 +1,382 @@
+                    ty,
+                    addr,
+                    mem_state,
+                    ..
+                },
+                &Node::Load {
+                    op: ref other_op,
+                    ty: other_ty,
+                    addr: other_addr,
+                    mem_state: other_mem_state,
+                    // Explicitly exclude: `inst` and `srcloc`. We
+                    // want loads to merge if identical in
+                    // opcode/offset, address expression, and last
+                    // store (this does implicit
+                    // redundant-load-elimination.)
+                    ..
+                },
+            ) => {
+                op == other_op
+                    && ty == other_ty
+                    && self.id_eq(addr, other_addr, uf)
+                    && mem_state == other_mem_state
+            }
+            _ => false,
+        }
+    }
+}
+
+impl CtxHash<Node> for NodeCtx {
+    fn ctx_hash(&self, value: &Node, uf: &mut UnionFind) -> u64 {
+        let mut state = crate::fx::FxHasher::default();
+        std::mem::discriminant(value).hash(&mut state);
+        match value {
+            &Node::Param {
+                block,
+                index,
+                ty: _,
+                loop_level: _,
+            } => {
+                block.hash(&mut state);
+                index.hash(&mut state);
+            }
+            &Node::Result {
+                value,
+                result,
+                ty: _,
+            } => {
+                self.hash_id(value, &mut state, uf);
+                result.hash(&mut state);
+            }
+            &Node::Pure {
+                ref op,
+                ref args,
+                types: _,
+            } => {
+                op.hash(&mut state);
+                self.hash_ids(args, &mut state, uf);
+                // Don't hash `types`: it requires an indirection
+                // (hence cache misses), and result type *should* be
+                // fully determined by op and args.
+            }
+            &Node::Inst { inst, ref args, .. } => {
+                inst.hash(&mut state);
+                self.hash_ids(args, &mut state, uf);
+            }
+            &Node::Load {
+                ref op,
+                ty,
+                addr,
+                mem_state,
+                ..
+            } => {
+                op.hash(&mut state);
+                ty.hash(&mut state);
+                self.hash_id(addr, &mut state, uf);
+                mem_state.hash(&mut state);
+            }
+        }
+
+        state.finish()
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct Cost(u32);
+impl Cost {
+    pub(crate) fn at_level(&self, loop_level: LoopLevel) -> Cost {
+        let loop_level = std::cmp::min(2, loop_level.level());
+        let multiplier = 1u32 << ((10 * loop_level) as u32);
+        Cost(self.0.saturating_mul(multiplier))
```

Comment:
I don't see the same sort of guard here in `at_level`, for saturating at less than "infinity", that I see in the implementation for `add`. Is that okay?

Reply:
Good point, fixed! I added the `finite()` helper here for this.